### PR TITLE
implement ptr::write without dedicated intrinsic

### PR DESCRIPTION
**Pull request by RalfJung**
This makes `ptr::write` more consistent with `ptr::write_unaligned`, `ptr::read`, `ptr::read_unaligned`, all of which are implemented in terms of `copy_nonoverlapping`.

This means we can also remove `move_val_init` implementations in codegen and Miri, and its special handling in the borrow checker.

Also see [this Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/ptr.3A.3Aread.20vs.20ptr.3A.3Awrite).

### CODE REVIEW

**Comment by RalfJung** [comment]
Let's see if perf says anything.

@bors try

**Comment by RalfJung** [comment]
@rust-timer queue

**Comment by RalfJung** [comment]
Let's see what happens if we only call intrinsics directly in `write`.

@bors try @rust-timer queue

**Comment by RalfJung** [comment]
Now perf is looking clean (but @bjorn3 remarked that the impact would likely be largest when *running* debug builds).

**Comment by bjorn3** [comment]
Perf results for https://github.com/ebobby/simple-raytracer/commit/804a7a21b9e673a482797aa289a18ed480e4d813:

```
Benchmark #1: ./raytracer_75e1acb
  Time (mean ± σ):      8.040 s ±  0.019 s    [User: 8.037 s, System: 0.004 s]
  Range (min … max):    8.013 s …  8.065 s    10 runs
 
Benchmark #2: ./raytracer_63c71d4
  Time (mean ± σ):      8.086 s ±  0.022 s    [User: 8.083 s, System: 0.003 s]
  Range (min … max):    8.061 s …  8.138 s    10 runs
 
Summary
  './raytracer_75e1acb' ran
    1.01 ± 0.00 times faster than './raytracer_63c71d4'
```

The perf difference is so small as simple-raytracer only spends a tiny bit of time in `std::ptr::write`. The codegen is noticably worse though.

Before:

```
       │    000000000007ab40 <core::ptr::write>:
       │    _ZN4core3ptr5write17h8616c78d24e53297E():
 32,01 │      sub    $0x18,%rsp
 19,97 │      mov    %rdi,0x8(%rsp)
 16,00 │      mov    %rsi,0x10(%rsp)
 32,02 │      mov    %rsi,(%rdi)
       │      add    $0x18,%rsp
       │    ← retq
```

After:

```
       │    000000000007ac00 <core::ptr::write>:
       │    _ZN4core3ptr5write17h8616c78d24e53297E():
 18,93 │      sub    $0x28,%rsp
  8,13 │      mov    %rsi,(%rsp)
 18,96 │      mov    %rdi,0x10(%rsp)
       │      movb   $0x0,0xf(%rsp)
  5,42 │      movb   $0x1,0xf(%rsp)
 10,72 │      mov    (%rsp),%rax
 29,71 │      mov    %rax,(%rdi)
  8,13 │      movb   $0x0,0xf(%rsp)
       │      add    $0x28,%rsp
       │    ← retq 
```

File: library/core/src/ptr/mod.rs
Comment (kpp) [inline library/core/src/ptr/mod.rs:891]: Is there a reason to use unsafe block in unsafe fn?

Comment (bjorn3) [reply library/core/src/ptr/mod.rs:891]: The `unsafe_op_in_unsafe_fn` is enabled in libcore.

**Comment by RalfJung** [comment]
@bjorn3 what kind of build of the raytracer is that (debug/release, which codegen backend)?

`copy_nonoverlapping` should directly lower to a `memcpy`. OTOH, looks like `move_val_init` is already lowered to an assignment during MIR building (but separate from @tmiasko's new pass). Looks like MIR assignments are lowered better than `copy_nonoverlapping`?

File: library/core/src/ptr/mod.rs
Comment (newpavlov) [inline library/core/src/ptr/mod.rs:886]: Add a comment on why it's linked again instead of using already imported `crate::intrinsics::copy_nonoverlapping` as done by other functions in this module?

**Comment by bjorn3** [comment]
> @bjorn3 what kind of build of the raytracer is that (debug/release, which codegen backend)?

Debug build using cg_llvm.

> copy_nonoverlapping should directly lower to a memcpy. OTOH, looks like move_val_init is already lowered to an assignment during MIR building (but separate from @tmiasko's new pass). Looks like MIR assignments are lowered better than copy_nonoverlapping?

Much better. `copy_nonoverlapping` results in several assignments and an extra taken reference in addition to the intrinsic call. It is even considered as capable of unwinding.

```rust
#![feature(core_intrinsics)]
pub unsafe fn write<T>(dst: *mut T, src: T) {
    std::intrinsics::move_val_init(&mut *dst, src)
}
```

```rust
fn write(_1: *mut T, _2: T) -> () {
    debug dst => _1;
    debug src => _2;
    let mut _0: ();
    let mut _3: *mut T;
    let mut _4: &mut T;

    bb0: {
        StorageLive(_4);
        _4 = &mut (*_1);
        _3 = &raw mut (*_4);
        (*_3) = move _2;
        StorageDead(_4);
        return;
    }
}
```

```rust
#![feature(core_intrinsics)]
pub unsafe fn write<T>(dst: *mut T, src: T) {
    std::intrinsics::copy_nonoverlapping(&src as *const T, dst, 1);
    std::intrinsics::forget(src);
}
```

```rust
fn write(_1: *mut T, _2: T) -> () {
    debug dst => _1;
    debug src => _2;
    let mut _0: ();
    let _3: ();
    let mut _4: *const T;
    let _5: &T;
    let mut _6: *mut T;
    let mut _7: bool;

    bb0: {
        _7 = const false;
        _7 = const true;
        StorageLive(_3);
        StorageLive(_4);
        StorageLive(_5);
        _5 = &_2;
        _4 = &raw const (*_5);
        StorageLive(_6);
        _6 = _1;
        _3 = copy_nonoverlapping::<T>(move _4, move _6, const 1_usize) -> [return: bb1, unwind: bb4];

    bb1: {
        StorageDead(_6);
        StorageDead(_4);
        StorageDead(_5);
        StorageDead(_3);
        _7 = const false;
        _0 = const ();
        return;
    }

    bb2 (cleanup): {
        resume;
    }

    bb3 (cleanup): {
        drop(_2) -> bb2;
    }

    bb4 (cleanup): {
        switchInt(_7) -> [false: bb2, otherwise: bb3];
    }
}
```

File: library/core/src/ptr/mod.rs
Comment (RalfJung) [reply library/core/src/ptr/mod.rs:886]: I will, if the reviewers agree that the PR should land at all.

The reason is to avoid the overhead of another function call indirection.

**Comment by RalfJung** [comment]
> Debug build using cg_llvm.

Oh I see. I guess the question is if we care, considering that this seems to only affect microbenchmarks.

> Much better. copy_nonoverlapping results in several assignments and an extra taken reference in addition to the intrinsic call. It is even considered as capable of unwinding.

That's on the MIR level; I was thinking of the LLVM level.

But potentially we could even lower some `copy_nonoverlapping` to assignments already on the MIR level?

**Comment by bjorn3** [comment]
> That's on the MIR level; I was thinking of the LLVM level.

In debug mode the mem2reg pass doesn't get run which I think is normally responsible for fixing these kinds of inefficiencies.

**Comment by m-ou-se** [comment]
This is now more a compiler change than a library change.

(The library changes in this PR looks fine to me.)

r? @lcnr

**Comment by lcnr** [comment]
I think that if removing an intrinsic without causing codegen regressions is a good thing.

Are the codegen issues already solved here? It isn't clear to me from just skimming over the discussion

File: library/core/src/ptr/mod.rs
Comment (usbalbin) [inline library/core/src/ptr/mod.rs:896]: What is the difference between `intrinsics::forget` and `mem::forget`? I saw that `mem::forget` is used by `write_unaligned` and `mem::forget` is `const fn`

Comment (bjorn3) [reply library/core/src/ptr/mod.rs:896]: `mem::forget` uses `intrinsics::forget` internally, but doesn't get inlined in debug mode.

Comment (usbalbin) [reply library/core/src/ptr/mod.rs:896]: Ah, ok. Thanks

Comment (RalfJung) [reply library/core/src/ptr/mod.rs:896]: Yeah, currently only `write` gets this special treatment.

We could also decide to make `copy_nonoverlapping` directly an intrinsic again without a wrapper function -- the debug assertions in there are disabled now anyway, and that would benefit all the functions in here, not just `write`.

**Comment by RalfJung** [comment]
@lcnr 
> Are the codegen issues already solved here? It isn't clear to me from just skimming over the discussion

There's a perf regression in debug builds. The question is whether that matters enough to block this cleanup PR.

File: library/core/src/ptr/mod.rs
Comment (lcnr) [inline library/core/src/ptr/mod.rs:886]:
```suggestion
    // We are calling the intrinsics directly to avoid function calls in the generated code
    // as `intrinsics::copy_nonoverlapping` is a wrapper function.
```

**Review by lcnr — APPROVED** [review]
the implementation looks good to me and from my experience we tend to mostly ignore debug performance.
So ignoring loss avoidance I don't think we would have added the `move_val_init` for that reason alone and I am in slightly in  favor of merging this PR.

@bjorn3 what's your opinion? Unless you - or someone else - strongly disagrees here, r=me once my review is resolved

**Comment by RalfJung** [comment]
The debug performance regression could likely be fixed by adding a post-drop-elaboration MIR optimization pass which replaces `copy_nonoverlapping(src, dst, 1)` by `*dst = *src`.

File: library/core/src/ptr/mod.rs
Comment (usbalbin) [reply library/core/src/ptr/mod.rs:897]: Am I right in assuming that we will want to make `write` `const` in a not too distant future? In that case is there anything in this PR to consider to prevent any future issues?

I believe borrowing `src` might be a problem in const context until insert_name_here is implemented. Do we think that is relevant to this PR?

Comment (usbalbin) [reply library/core/src/ptr/mod.rs:897]: > I believe borrowing src might be a problem in const context until insert_name_here is implemented

On second thought, is that #80418? :)

Comment (RalfJung) [reply library/core/src/ptr/mod.rs:897]: While yes, we want to make it `const`, I also don't think we should make our runtime implementation worse, or keep such a complex intrinsic around, just to be able to make it `const`.

But also, as you said, with https://github.com/rust-lang/rust/pull/80418 this function should actually be fine.

Comment (usbalbin) [reply library/core/src/ptr/mod.rs:897]: > While yes, we want to make it const, **I also don't think we should make our runtime implementation worse, or keep such a complex intrinsic around, just to be able to make it const**.

Not what what I meant to suggest :)

> But also, as you said, with #80418 this function should actually be fine.

Awesome then :)

**Comment by RalfJung** [comment]
@lcnr how do we proceed with this?

**Comment by lcnr** [comment]
let's merge this with the rationale provided in https://github.com/rust-lang/rust/pull/80290#pullrequestreview-560641762

@bors r+ rollup=never

**Comment by therealprof** [comment]
There's a noticeable binary size regression between `rustc 1.51.0-nightly (e38fb306b 2021-01-14)` and `rustc 1.51.0-nightly (4253153db 2021-01-17)` on `thumbv6m-none-eabi` in dev mode. However looking at the assembly I have troubles accounting for all of it:
```
Section size:
   text    data     bss     dec     hex filename
   9820       0       4    9824    2660 blinky
```
vs
```
Section size:
   text    data     bss     dec     hex filename
   9884       0       4    9888    26a0 blinky
```

The only observable difference I can spot is:
```
 08000936 <core::ptr::write>:
- 8000936:      b083            sub     sp, #12
- 8000938:      9001            str     r0, [sp, #4]
- 800093a:      9102            str     r1, [sp, #8]
- 800093c:      6001            str     r1, [r0, #0]
- 800093e:      b003            add     sp, #12
- 8000940:      4770            bx      lr
+ 8000936:      b082            sub     sp, #8
+ 8000938:      9100            str     r1, [sp, #0]
+ 800093a:      9001            str     r0, [sp, #4]
+ 800093c:      9900            ldr     r1, [sp, #0]
+ 800093e:      6001            str     r1, [r0, #0]
+ 8000940:      b002            add     sp, #8
+ 8000942:      4770            bx      lr
```

**Comment by RalfJung** [comment]
@therealprof so is the binary of rustc itself bigger, or is some rustc-generated binary bigger? Is this a debug build or a release build?

https://github.com/kennytm/rustup-toolchain-install-master could be used to confirm that it is this PR vs some other PR that landed that day.

**Comment by therealprof** [comment]
> @therealprof so is the binary of rustc itself bigger, or is some rustc-generated binary bigger? Is this a debug build or a release build?

Generated binaries in dev (or debug mode if you prefer) are larger.

> https://github.com/kennytm/rustup-toolchain-install-master could be used to confirm that it is this PR vs some other PR that landed that day.

Sorry, I don't have time to bisect this.

I was just browsing the recently merged PRs and the mention of regressions piqued my interest so I decided to run my tools (https://github.com/stm32-rs/stm32f0xx-hal/blob/master/tools/capture_nightly_example_bloat.sh) on the latest nightly and sure enough I can see regressions happening between `rustc 1.51.0-nightly (e38fb306b 2021-01-14)` and `rustc 1.51.0-nightly (4253153db 2021-01-17)`.

**Comment by RalfJung** [comment]
Some regression for debug builds was expected with this PR. Depending on how much this matters, I sketched an idea for how to mitigate this:

> The debug performance regression could likely be fixed by adding a post-drop-elaboration MIR optimization pass which replaces copy_nonoverlapping(src, dst, 1) by *dst = *src.

**Comment by therealprof** [comment]
Well, debug mode in Rust is atrocious in every regard compared to other languages used for embedded development. The binaries are huge (often too big to fit into the flash of smaller microcontrollers) and slow (some peripherals like USB can't even be used since the generated code is some order of magnitudes too slow to react to USB events).

Every little improvement helps! I'd be happy to test and benchmark any changes once they've landed in nightly. I've collected a nice dataset over different versions in the current format (each stable release back to 1.41 and a quite a few more nightlies in between).

**Comment by RalfJung** [comment]
The reason this PR was deemed acceptable is that the issue should only affect debug builds. Anybody who cares about such issues would use release (or size-optimized) builds, we figured, and those should not be affected.

Why does the size of debug builds matter so much to you?

(Btw, discussions in a closed PR are bound to get lost, so if you think something should be done here or if you think there's something worth tracking, please open an issue.)

> Every little improvement helps! I'd be happy to test and benchmark any changes once they've landed in nightly. 

I won't have time to work on this, but maybe someone from @rust-lang/wg-mir-opt would be interesting. Cc @tmiasko who recently added a closely related intrinsic lowering MIR pass.

**Comment by therealprof** [comment]
> The reason this PR was deemed acceptable is that the issue should only affect debug builds. Anybody who cares about such issues would use release (or size-optimized) builds, we figured, and those should not be affected.

I can confirm release builds are not affected.

> Why does the size of debug builds matter so much to you?

Well, only debug builds are really debuggable (in embedded context) for starters and debugging with a debugger has some extra relevance due to limited interaction capabilities compared with a regular application. Release builds are also built with all optimisation features Rust has to offer which really make them really slow to compile. And then there's the usual trap for young players that the default build mode is dev mode, so starters are frequently ending up in debug mode; best case they'll get a linker error telling them about their mistake...

> (Btw, discussions in a closed PR are bound to get lost, so if you think something should be done here or if you think there's something worth tracking, please open an issue.)

Sure thing.

**Comment by rylev** [comment]
@RalfJung @lcnr There seems to be a [small compile time perf regression after all in regex-debug full builds](https://perf.rust-lang.org/compare.html?start=63a83c5f55801b17b77adf690db397d17c706c48&end=492b83c6971c390af7a42932869502224ef55ef7&stat=instructions:u). Looks like LLVM_module_codegen_emit_obj regressed.Thoughts on if addressing this is worth it?

**Comment by oli-obk** [comment]
Considering that `LLVM_module_optimize_module_passes` now takes half the time, I think LLVM now fails to optimize something that then causes more llvm object code to be emitted. This could also have a runtime impact due to missing some optimizations.

Now... this is only for regex-debug, and we generally don't put too much effort into making debug builds be efficient, so the runtime perf loss is acceptable, but it is still not great that we lost some compile-time. I guess regex uses `ptr::write` a lot (or at least "for a lot of different `T`").

I don't think it is possible to regain this perf directly. We may get it back via always-run MIR optimizations that turn this `ptr::write` into basic MIR statements.

**Comment by therealprof** [comment]
cf. https://github.com/rust-lang/rust/issues/81163 for ongoing discussion

> Now... this is only for regex-debug, and we generally don't put too much effort into making debug builds be efficient, so the runtime perf loss is acceptable,

I very much disagree with this statement. The performance and binary size of debug builds are a huge problem for embedded Rust already; regressions are definitely not acceptable for us.

**Comment by RalfJung** [comment]
Also see https://github.com/rust-lang/rust/pull/80290#issuecomment-753490300 for an approach that should regain perf. `write` wasn't itself an intrinsic before, so I don't think it is necessary to make it an intrinsic now.

**Comment by rylev** [comment]
Should we create an issue for https://github.com/rust-lang/rust/pull/80290#issuecomment-763583590 to make sure it's not lost?

**Comment by RalfJung** [comment]
Yeah if people care enough, there should be a version of https://github.com/rust-lang/rust/issues/81163 for `ptr::write`.

**Pull request merged**
